### PR TITLE
Doctrine DBAL 2.13 blog post

### PR DESCRIPTION
**Pull request by beberlei**

### CODE REVIEW

File: source/blog/2021-03-29-dbal-2.13.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
+---
+title: "New Release: Doctrine DBAL 2.13 with Deprecations and Forward Compatibility"
+authorName: Benjamin Eberlei
+authorEmail: kontakt@beberlei.de
+permalink: /2021/03/29/dbal-2.13.html
+---
+
+We have released the last minor version in the 2.x family of Doctrine DBAL with
+additional forward compatibilty to DBAL 3 around Statement and Result classes
```
Comment (greg0ire) [inline source/blog/2021-03-29-dbal-2.13.md:9]:
```suggestion
additional forward compatibilty to DBAL 3 around `Statement` and `Result` classes
```

File: source/blog/2021-03-29-dbal-2.13.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+$stmt->fetchAllAssociative();
+```
+
+Many more changes have been made on the public API and also for the internals,
+but these are the most visible ones.
+
+### Deprecations Logging
+
+We have wrestled internally for a long time with the strategy on runtime
+deprecation going forward and settled on a introducing a tiny abstraction on
```
Comment (greg0ire) [inline source/blog/2021-03-29-dbal-2.13.md:66]:
```suggestion
deprecation going forward and settled on introducing a tiny abstraction on
```

File: source/blog/2021-03-29-dbal-2.13.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+The reason for this abstraction is the potential for side effects caused by an
+error handler and the potential overhead. We expect our deprecations to be
+triggered a few hundred times in some requests as such the production overhead
+must be minimal. 
+
+This means deprecation logging is disabled by default and you must enable
+it to either use `@trigger_error` or a PSR-3 compatible logger.
+
+See the [deprecation library
+README.md](https://github.com/doctrine/deprecations/) for details how to
```
Comment (greg0ire) [inline source/blog/2021-03-29-dbal-2.13.md:78]:
```suggestion
README.md](https://github.com/doctrine/deprecations/) for details on how to
```

File: source/blog/2021-03-29-dbal-2.13.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+both DBAL 2 or 3 this version of Doctrine DBAL will work again with PHP 7.1 and
+7.2
+
+### Migrate to DBAL 3
+
+We recommend a three step strategy to move your code-base and that of your
+dependents to DBAL 3:
+
+1. Upgrade to DBAL 2.13 and enforce "^2.13" as a version constraint in
+   Composer.json
```
Comment (greg0ire) [inline source/blog/2021-03-29-dbal-2.13.md:97]:
```suggestion
   composer.json
```

File: source/blog/2021-03-29-dbal-2.13.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+2. Enable deprecation tracking and eliminate all deprecations triggered in your
+   codebase. If you have downstream packages that use your application
+   or library, provide them with a release that depends just on DBAL 2.13 
+3. After fixing all deprecations, update composer constraint to "^2.13 | ^3.0".
```
Comment (greg0ire) [inline source/blog/2021-03-29-dbal-2.13.md:101]:
```suggestion
3. After fixing all deprecations, update Composer constraint to "^2.13 | ^3.0".
```

File: source/blog/2021-03-29-dbal-2.13.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+   Doctrine will upgrade to version 3 once all other dependencies you are using
+   are ready as well to upgrade to version 3.
+
+We recommend the following strategies to detect the use of deprecated code:
+
+1. If you have an extensive test-suite, register a PSR-3 logger with Doctrine
+   Deprecations to catch all deprecations while running the tests.
+2. Otherwise register a PSR-3 logger with Doctrine Deprecations in development
+   or staging only and collect and fix them as good you can. This could be
```
Comment (greg0ire) [inline source/blog/2021-03-29-dbal-2.13.md:110]:
```suggestion
   or staging only and collect and fix them as well you can. This could be
```

File: source/blog/2021-03-29-dbal-2.13.md (hunk 4 above)
Comment (morozov) [reply source/blog/2021-03-29-dbal-2.13.md:109]: This is not addressed.

File: source/blog/2021-03-29-dbal-2.13.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
+---
+title: "New Release: Doctrine DBAL 2.13 with Deprecations and Forward Compatibility"
+authorName: Benjamin Eberlei
+authorEmail: kontakt@beberlei.de
+permalink: /2021/03/29/dbal-2.13.html
+---
+
+We have released the last minor version in the 2.x family of Doctrine DBAL with
```
Comment (morozov) [inline source/blog/2021-03-29-dbal-2.13.md:8]: Never say never. I'd rephrase it with something like "there's no plan to release minor versions in the 2.x series". I believe `2.11` was the first "last" minor and we communicated it as "last planned".

File: source/blog/2021-03-29-dbal-2.13.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+additional forward compatibilty to DBAL 3 around `Statement` and `Result` classes
```
Comment (morozov) [inline source/blog/2021-03-29-dbal-2.13.md:9]: Technically, there are both classes and `interfaces` named `Statement` and `Result` that this article refers to. I'd be less specific and refer to them as APIs.

File: source/blog/2021-03-29-dbal-2.13.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+and with an integration in our new deprecations logging library.
+
+In addition this DBAL release re-enables PHP 7.1 and 7.2 compatibility to give
+as much flexibility as possible to everyone with forward compatibility.
+
+### Statement and Result Forward Compatibility
+
+DBAL 3.0 extracts all fetch-methods from the `Statement` class and moved them
```
Comment (morozov) [inline source/blog/2021-03-29-dbal-2.13.md:18]: Same as above, it's not only classes but interfaces as well.

File: source/blog/2021-03-29-dbal-2.13.md [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+DBAL 2 and 3 at the same time.
+
+Old code:
+
+```php
+$statement = $connection->prepare('SELECT * FROM tbl WHERE col = ?');
+$statement->bindParam(1, $value);
+$statement->execute();
+
+while ($row = $statement->fetch()) {
```
Comment (morozov) [inline source/blog/2021-03-29-dbal-2.13.md:30]: I know it's just an example but we've had a few iteration-related bugs in the DBAL where an empty value fetched from a recordset would break iteration. It should be `while (($row = $statement->fetch()) !== false)` to not set a bad example.

File: source/blog/2021-03-29-dbal-2.13.md [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+```
+
+New Code:
+
+```php
+$statement = $connection->prepare('SELECT * FROM tbl WHERE col = ?');
+$statement->bindParam(1, $value);
+$result = $statement->execute();
+
+while ($row = $result->fetchAssociative()) {
```
Comment (morozov) [inline source/blog/2021-03-29-dbal-2.13.md:43]: Should compare strictly to `false`.

File: source/blog/2021-03-29-dbal-2.13.md [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+but these are the most common ones.
+
+Thank you to [mdumoulin](https://github.com/mdumoulin) for the work on
+improving the forward compatiblity.
+
+### Deprecations Logging
+
+We have wrestled internally for a long time with the strategy on runtime
+deprecation going forward and settled on introducing a tiny abstraction on
+top of `@trigger_error()`.
```
Comment (morozov) [inline source/blog/2021-03-29-dbal-2.13.md:77]: I think the whole point of this abstraction is that it doesn't unconditionally call `@trigger_error()`. I'd rephrase it to "an API for logging the usage of deprecated APIs". Something like that.

File: source/blog/2021-03-29-dbal-2.13.md [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+are not officially supported anymore, but are covered by support of a few Linux
+distributions.
+
+To provide the largest possible flexibility to the ecosystem to run code with
+both DBAL 2 or 3 this version of Doctrine DBAL will work again with PHP 7.1 and
+7.2
+
+### Migrate to DBAL 3
+
+We recommend a three step strategy to move your code-base and that of your
```
Comment (morozov) [inline source/blog/2021-03-29-dbal-2.13.md:103]: Is it "a three-step strategy" or there are two three-step strategies? The existing wording of the intro slightly contradicts the body of the paragraph.

File: source/blog/2021-03-29-dbal-2.13.md [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+We recommend the following strategies to detect the use of deprecated code:
+
+1. If you have an extensive test-suite, register a PSR-3 logger with Doctrine
+   Deprecations to catch all deprecations while running the tests.
+2. Otherwise register a PSR-3 logger with Doctrine Deprecations in development
+   or staging only and collect and fix them as well you can. This could be
+   done for a longer amount of time of days, weeks or months.
+3. If that is not possible, register a PSR-3 logger in production. Make sure to
+   eliminate high frequency deprecations quickly or call `ignoreDeprecation` to
+   snooze them to avoid overhead.
```
Comment (morozov) [inline source/blog/2021-03-29-dbal-2.13.md:138]: It may be worth mentioning Psalm which reports the usage of deprecated symbols at level 2. Not sure if it's adopted at this level by the consumers of the DBAL but it's one of the most reliable options (doesn't depend on test coverage, doesn't require any changes in production).

File: source/blog/2021-03-29-dbal-2.13.md [hunk 15]
```diff
@@ -128,15 +128,12 @@ For a library, framework or platform:
...
-   or staging only and collect and fix them as well you can. This could be
-   done for a longer amount of time of days, weeks or months.
-3. If that is not possible, register a PSR-3 logger in production. Make sure to
-   eliminate high frequency deprecations quickly or call `ignoreDeprecation` to
-   snooze them to avoid overhead.
-
-When you are part of the Symfony ecosystem and use the symfony/error-handler
-component, then you can also use the `trigger_error` mechanism instead of PSR-3
-logging.
+- Use Psalm or other static analyers to detect the use of Deprecated code.
```
Comment (morozov) [inline source/blog/2021-03-29-dbal-2.13.md:131]: There's a typo in "analyers". "Deprecated" shouldn't be capitalized. We can mention IDEs here as well.

File: source/blog/2021-03-29-dbal-2.13.md [hunk 16]
```diff
@@ -128,15 +128,12 @@ For a library, framework or platform:
...
+- If you have an extensive test-suite, register a PSR-3 logger with Doctrine
+  Deprecations to catch all deprecations while running the tests.
+- Otherwise register a PSR-3 logger with Doctrine Deprecations in development
+  or staging only and collect and fix them as well you can. This could be
+  done for a longer amount of time of days, weeks or months.
+- If that is not possible, register a PSR-3 logger in production. Make sure to
```
Comment (morozov) [inline source/blog/2021-03-29-dbal-2.13.md:137]: What would be a case where one cannot enable a PSR-3 logger in dev or stage but can in production?

Comment (beberlei) [reply source/blog/2021-03-29-dbal-2.13.md:137]: not having a stage

File: source/blog/2021-03-29-dbal-2.13.md [hunk 17]
```diff
@@ -5,9 +5,10 @@ authorEmail: kontakt@beberlei.de
 permalink: /2021/03/29/dbal-2.13.html
 ---
 
-We have released the last minor version in the 2.x family of Doctrine DBAL with
-additional forward compatibilty to DBAL 3 around `Statement` and `Result` classes
-and with an integration in our new deprecations logging library.
+We have released DBAL 2.13, what we plan to be the last minor version in the 2.x family of Doctrine DBAL.
+This release includes additional forward compatibilty to DBAL 3 around
```
Comment (greg0ire) [inline source/blog/2021-03-29-dbal-2.13.md:9]:
```suggestion
This release includes additional forward compatibility to DBAL 3 around
```
